source/_posts/2015-11-16-immutable-strings-in-ruby-2-dot-3.markdown

```diff
@@ -0,0 +1,279 @@
+an error will be thrown. Therefore race conditions are gone.
+
+Certainly forgetting about race conditions in certain scenarios is great. But
+(there's always a but) bear in mind that string literals are thread safe, not
+its references!
+
+{% codeblock lang:ruby %}
+class Foo
+  attr_accesor :literal
+
+  def initialize(literal)
+    @literal = literal
+  end
+end
+{% endcodeblock %}
+
+Picture yourself working with a `Foo` class, what happens when we pass a `Foo`
+object to several threads, some of them reading from `literal` attribute and
+another ones writing to?
+
+In this case despite we are using string literals, we aren't safe when it comes
+to thread safety. Actually we're holding a reference to a literal, nothing stop
+us from changing that reference for another literal.
+
+# Benchmarks
+
+In this last section we're going to run a series of benchmarks to see how
+strings immutability impacts on object allocation, garbage collection and
+performance. Let's take the following snippet as the starting point for the
+benchmarks.
+
+{% codeblock lang:ruby %}
+require 'objspace'
+
+GC.start
+GC.disable
+
+1_000_000.times {}
+
+puts ObjectSpace.count_objects[:T_STRING]
+{% endcodeblock %}
+
+Running the previous snippet creates 5,328 and 5,329 objects with and without
+frozen strings enabled respectively. Let's change `1_000_000.times {}` for
+`1_000_000.times { 'a' }` and run the script again: it returns 5,329 and
+1,005,330. The frozen string version differs only by one object (the extra `a`)
+meanwhile without frozen strings an extra million objects are being created.
+
+Now let's enable garbage collection, remove object count and see what `GC.stat`
+has to say. `GC.stat` will return a hash with garbage collection stats, the key we are
+interested on are: `:count` and `:total_allocated_objects`.
+
+{% codeblock lang:ruby %}
+GC.start
+1_000_000.times {}
+puts GC.stat
+{% endcodeblock %}
+
+
+This is the formatted output:
+
+
+Version|GC runs|Object allocation
+-|-|-
+mutable|57|1,057,077
+immutable|6|57078
+
+Finally let's measure how this impacts on performance.
+
+
+{% codeblock lang:ruby %}
+require 'benchmark'
+
+puts Benchmark.measure { 1_000_000.times { 'a' } }
+{% endcodeblock %}
+
+On the table below you can find the results obtained from running the benchmark
+without and with frozen literals respectively.
+
+<table>
+  <thead>
+    <tr>
+      <th>User</th>
+      <th>System</th>
+      <th>Total</th>
+      <th>Real</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>0.080000 </td>
+      <td>0.000000</td>
+      <td>0.080000</td>
+      <td>0.073067</td>
+    </tr>
+    <tr>
+      <td>0.040000</td>
+      <td>0.000000</td>
+      <td>0.040000</td>
+      <td>0.045621</td>
+    </tr>
+  </tbody>
+</table>
+
+
+As you can see running the benchmark with frozen string literal result in a
+speed improvement of 37%. The main responsible for this is what we saw on the
+previous benchmarks: allocated objects.
```

Comment:
I'd explain something like ... `this shows an improvement only for this little example. We can't conclude a lot about web applications from this`